src/software/thunderscope/replay/proto_player.py

```diff
@@ -0,0 +1,334 @@
+            else:
+                self.pause()
+
+    def set_playback_speed(self, speed):
+        """Sets the playback speed.
+
+        :param speed: The speed to set the playback to.
+
+        """
+        with self.replay_controls_mutex:
+            self.pause()
+            self.playback_speed = 1.0 / float(speed)
+            self.play()
+
+    def seek(self, seek_time):
+        """Seeks to a specific time. We binary search through the chunks
+        to find the chunk that would contain the data at the given time.
+
+        We then binary search the entries in the chunk to find the entry
+        cloest to the given time.
+
+        We then set the current_chunk_index and current_entry_index to
+        the correct values. We also update the seek_offset_time to help
+        with realtime playback timing calculations in the worker thread.
+
+        :param seek_time: The time to seek to.
+
+        """
+        # Lets binary search through the chunks to find the chunk that starts
+        # with a timestamp less than (but closest to) the seek_time we want
+        # to seek to.
+        def __bisect_chunks_by_timestamp(chunk):
+            chunk = ProtoPlayer.load_replay_chunk(chunk)
+            start_timestamp, _, _ = ProtoPlayer.unpack_log_entry(chunk[0])
+            return start_timestamp
+
+        with self.replay_controls_mutex:
+            self.current_chunk_index = ProtoPlayer.binary_search(
+                self.sorted_chunks, seek_time, key=__bisect_chunks_by_timestamp
+            )
+
+        # Lets binary search through the entries in the chunk to find the closest
+        # timestamp to seek to
+        def __bisect_entries_by_timestamp(entry):
+            timestamp, _, _ = ProtoPlayer.unpack_log_entry(entry)
+            return timestamp
+
+        with self.replay_controls_mutex:
+
+            # Load the chunk that would have the entry
+            self.current_chunk = ProtoPlayer.load_replay_chunk(
+                self.sorted_chunks[self.current_chunk_index]
+            )
+
+            # Search through the chunk to find the entry that is closest to
+            # the seek_time
+            self.current_entry_index = ProtoPlayer.binary_search(
+                self.current_chunk, seek_time, key=__bisect_entries_by_timestamp
+            )
+
+            # Update the seek_offset_time and current_packet_time
+            # to the one we just found.
+            self.seek_offset_time = __bisect_entries_by_timestamp(
+                self.current_chunk[self.current_entry_index]
+            )
+            self.current_packet_time = self.seek_offset_time
+
+            logging.info(
+                "Jumped to chunk {} at index {} with timestamp {:.2f}".format(
+                    self.current_chunk_index,
+                    self.current_entry_index,
+                    self.seek_offset_time,
+                )
+            )
+
+    @staticmethod
+    def binary_search(arr, x, key=lambda x: x):
```

Comment:
there's also this: https://stackoverflow.com/questions/212358/binary-search-bisection-in-python

Reply:
I looked into this before grabbing the custom soln, its only available in python3.10 which is more than an apt-get away (you have to add a custom repo and then download python3.10, its not readily available on ubuntu 20 rn AFAICT)

Reply:
:o